### PR TITLE
Bug 2034460: cloud-provider: handle advanced AWS and Azure configurations 

### PR DESCRIPTION
**Pull request by squeed**
It turns out connecting to AWS and Azure isn't so simple. This change
does the following:

- Reads the AWS endpoint override from the Infrastructure object
- Pulls in the AWS CA bundle from the kube-cloud-config ConfigMap
- Reads the Azure Environment and URL overrides from the Infrastructure
  object

Fixes: 2034460

### CODE REVIEW

File: pkg/cloudprovider/cloudprovider.go
Comment (alexanderConstantinescu) [inline pkg/cloudprovider/cloudprovider.go:113]: We can't do it like this. There can be no hard dependency from the CNCC on other OpenShift components, this is because upstream OVN-Kubernetes needs to be able to deploy this component and have it work by only providing the secrets required to talk to the cloud API, as mentioned in the enhancement proposal: https://github.com/alexanderConstantinescu/enhancements/blob/3231f1103145f5ef5b7b88bc6aab811b834088b0/enhancements/network/cloud-egress-ip.md#openshift-vs-upstream-project . Upstream should not to deploy other OpenShift components to have this one work. Hence we need a default behavior intended for reading "simple credenentials/secrets" and then override that in case we deploy something "obscure only OpenShift does" type of thing.

Comment (alexanderConstantinescu) [reply pkg/cloudprovider/cloudprovider.go:113]: This is also the reason the `platform-type` / `platform-region` etc is passed to this component though its deployment stanza

**Review by alexanderConstantinescu — CHANGES_REQUESTED** [review]

**Comment by squeed** [comment]
OK, updated to just get configuration from cli args. Will need a CNO PR too, will file shortly.

**Comment by squeed** [comment]
@alexanderConstantinescu PTAL, thanks

File: pkg/cloudprovider/azure.go
Comment (alexanderConstantinescu) [inline pkg/cloudprovider/azure.go:72]: So there always needs to be either an `APIOverride` or `AzureEnvironment` ?

File: cmd/cloud-network-config-controller/main.go
Comment (alexanderConstantinescu) [inline cmd/cloud-network-config-controller/main.go:177]: Nit:  uppercase "api" in the description.

**Review by alexanderConstantinescu — COMMENTED** [review]
/hold

Until the CNO PR is merged. 

Besides that I only have some minor comments, this looks good in general. Thanks!

File: pkg/cloudprovider/azure.go
Comment (squeed) [reply pkg/cloudprovider/azure.go:72]: Kind of, there is a default Azure environment specified. But Azure makes liberal use of the "environment" -- you even need to change it if you're using the Azure Germany cloud.

**Comment by squeed** [comment]
/hold cancel
This needs to merge first, then the CNO change can go in (so it can use the new cli flags)

**Comment by alexanderConstantinescu** [comment]
> This needs to merge first, then the CNO change can go in (so it can use the new cli flags)

Is it though? `AzureEnvironment` or `APIOverride` are required on Azure as things stand with this PR...yet the CNO doesn't provide these arguments. We will break all OpenShift CI jobs on Azure if we merge this before the CNO one, right?

**Comment by alexanderConstantinescu** [comment]
It's required given this https://github.com/openshift/cloud-network-config-controller/pull/14#discussion_r781239078 right?

This would make it "not required"

```go
        if a.cfg.APIOverride != "" {
		a.env, err = azure.EnvironmentFromURL(a.cfg.APIOverride)
	} else if a.cfg.AzureEnvironment != "" {
		a.env, err = azure.EnvironmentFromName(a.cfg.AzureEnvironment)
	}
```

Well, I am assuming `azure.EnvironmentFromName` will fail for empty strings, maybe that is not the case?

I don't see any problems with AWS, so it should probably pass this repo's CI.

**Comment by squeed** [comment]
huh, terraform wierdness:

```
level=error msg=When applying changes to module.vpc.aws_subnet.public_subnet[1], provider
level=error msg="openshift/local/aws" produced an unexpected new value: Root resource was
level=error msg=present, but now absent.
level=error
level=error msg=This is a bug in the provider, which should be reported in the provider's own
level=error msg=issue tracker. 
```

/retest

**Comment by squeed** [comment]
/hold
can't get Azure to come up w/ cluster-bot

**Comment by squeed** [comment]
argh
> level=error
level=error msg=Error: creating/updating Public Ip Address: (Name "ci-ln-m41tvj2-1d09d-4cr94-bootstrap-pip-v4" / Resource Group "ci-ln-m41tvj2-1d09d-4cr94-rg"): network.PublicIPAddressesClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="PublicIPCountLimitReached" Message="Cannot create more than 100 public IP addresses for this subscription in this region." Details=[]
level=error

**Comment by squeed** [comment]
/hold cancel
azure works fine

**Review by alexanderConstantinescu — APPROVED** [review]
/lgtm

**Pull request merged**
